### ISSUE TITLE
Movim ignores carbons for incoming messages

### ISSUE THREAD

**Issue by dos1** [movim/movim#1182]
Haven't managed to debug or even reproduce reliably yet, so unfortunately I don't have much to offer in this description so far, but there's definitely something wrong with history retrieval after logging in. I'm mostly using other clients myself, but when I log in to my account via Movim, I occasionally see the chat history since my last login to Movim consisting only of my outgoing messages and completely missing the incoming ones. The missing messages aren't present in the database at all.

The history is correctly retrieved by other clients such as Dino, Chatty or Gajim. This is happening with server running ejabberd 21.01 from Debian Bullseye with:
```
  mod_mam:
    default: always
    assume_mam_usage: true
```

**Comment by ticktostick** [comment]
same here 
very buggy
not getting all message history in movim

other clients works ok (conversations and gajim)

**Comment by edhelas** [comment]
A big refactoring regarding MAM was done a few months ago and a release was done since then.

**Comment by dos1** [comment]
I think it was before v0.24 was tagged, right? If so, then this wasn't fixed, as I could still see this happening in v0.24.

Though I see some MAM fixes since v0.24 as well. I'm updating right now, will keep an eye on it.

**Comment by dos1** [comment]
Nope, still happening on v0.27.1.

**Comment by edhelas** [comment]
To handle this ticket I'd need more information and proper way to reproduce it.

What are exactly the issues ? What is not working, which messages from the history are not loaded ? 
Having some logs and a "tutorial" would greatly help. If not I'll not be able to do better than what's now.

**Comment by dos1** [comment]
It's easiest to notice when using another client while away from Movim. It happens quite often in fact, but I've never caught it on demand so far, so I can't offer instructions to reproduce it yet. You can see one side of the conversation - carbons from your other client - but not the other side, so the conversation in Movim looks like you were talking to yourself. Nothing relevant or interesting in the logs.

How does Movim handle sessions? How long is XMPP session supposed to linger after the user disconnects from Movim's websocket? I'm not quite sure this is related to MAM.

**Comment by dos1** [comment]
Just experienced some messages missing. Movim's XMPP session was active, but its presence was set to away. I believe that's the problematic state; I can receive messages reliably when Movim is open or when I'm logged out of Movim (destroying the XMPP session), but not after closing Movim, leaving the session active and waiting for its presence to be set to away.

**Comment by dos1** [comment]
Not sure what took me so long to figure this out, it turns out to be really simple: it's just carbons.

Movim completely ignores carbons about incoming messages received by other clients. The message is received by the daemon, but ignored:

```xml
<message to='dos@dosowisko.net/movim5Fn1R6' from='dos@dosowisko.net' type='chat'>
    <received xmlns='urn:xmpp:carbons:2'>
        <forwarded xmlns='urn:xmpp:forward:0'>
            <message xml:lang='en' to='dos@dosowisko.net/dino.2c5bef18' from='test@dosowisko.net/pliszka' type='chat' id='e9d7f052-e470-45fa-a0c5-794667ebd7d2'
                xmlns='jabber:client'>
                <archived by='dos@dosowisko.net' id='1725927508262418'
                    xmlns='urn:xmpp:mam:tmp'/>
                <stanza-id by='dos@dosowisko.net' id='1725927508262418'
                    xmlns='urn:xmpp:sid:0'/>
                <active xmlns='http://jabber.org/protocol/chatstates'/>
                <request xmlns='urn:xmpp:receipts'/>
                <origin-id xmlns='urn:xmpp:sid:0' id='e9d7f052-e470-45fa-a0c5-794667ebd7d2'/>
                <body>dino</body>
            </message>
        </forwarded>
    </received>
</message>
<message to='dos@dosowisko.net/movim5Fn1R6' from='dos@dosowisko.net'>
    <sent xmlns='urn:xmpp:carbons:2'>
        <forwarded xmlns='urn:xmpp:forward:0'>
            <message xml:lang='en' to='test@dosowisko.net/pliszka' from='dos@dosowisko.net/dino.2c5bef18' id='93551cd0-efa7-407c-ba60-6cf32b5dcec2'
                xmlns='jabber:client'>
                <received xmlns='urn:xmpp:receipts' id='e9d7f052-e470-45fa-a0c5-794667ebd7d2'/>
            </message>
        </forwarded>
    </sent>
</message>
<message to='dos@dosowisko.net/movim5Fn1R6' from='dos@dosowisko.net'>
    <sent xmlns='urn:xmpp:carbons:2'>
        <forwarded xmlns='urn:xmpp:forward:0'>
            <message xml:lang='en' to='test@dosowisko.net/pliszka' from='dos@dosowisko.net/1617532562767235691719891'
                xmlns='jabber:client'>
                <received xmlns='urn:xmpp:receipts' id='e9d7f052-e470-45fa-a0c5-794667ebd7d2'/>
            </message>
        </forwarded>
    </sent>
</message>
```

Outgoing carbons show up correctly.

**Comment by edhelas** [comment]
So far I cannot reproduce the issue, the carbon messages are handled properly by Movim both way, online and with MAM from the tests I did. 

Can you please give me the few step to reproduce (connect on Dino with A, send a message to B, connect on Movim...).

**Comment by dos1** [comment]
I wasn't able to get any incoming carbon to show up in my tests, only outgoing ones. The only case where messages sent to bare JID arrived was when Movim's presence had the highest priority or shared the same priority with other clients, in which case the messages are received directly without carbons.

So the reproducer is simply:
1. Connect to Movim.
2. Connect to the same account with another client with higher presence priority.
3. Observe incoming messages not being displayed in Movim despite of being received from the server as carbons.

I just reproduced it on mov.im with zero troubles.

Ignore all the other stuff I mentioned earlier about idling sessions etc, it was caused by priority being changed to -1 when idle on my installation. The real issue is just about carbons.
